File: system-design/oop-patterns/03-decorator-pattern/full/js/main.js
import { Espresso } from './coffee/Espresso.js';
import { Caramel } from './coffee/Caramel.js';
import { Soy } from './coffee/Soy.js';

const espresso = new Espresso();
const caramel = new Caramel(espresso);
const soy = new Soy(caramel);

console.log('Espresso:', espresso.cost());
console.log('Espresso + Caramel:', caramel.cost());
console.log('Espresso + Caramel + Soy:', soy.cost());

File: system-design/oop-patterns/03-decorator-pattern/full/cpp/main.cpp
#include <iostream>
#include "coffee/Espresso.h"
#include "coffee/Caramel.h"
#include "coffee/Soy.h"

int main() {
    Espresso espresso;
    Caramel caramel(&espresso);
    Soy soy(&caramel);

    std::cout << "Espresso: " << espresso.cost() << "\n";
    std::cout << "Espresso + Caramel: " << caramel.cost() << "\n";
    std::cout << "Espresso + Caramel + Soy: " << soy.cost() << "\n";

    return 0;
}
